block/internal/syncing/syncer.go

```diff
@@ -252,61 +267,94 @@ func (s *Syncer) processLoop() {
 }
 
 func (s *Syncer) startSyncWorkers() {
-	s.wg.Add(3)
+	s.wg.Add(2)
 	go s.daWorker()
-	go s.p2pWorker()
 	go s.pendingWorker()
 }
 
-func (s *Syncer) daWorker() {
-	defer s.wg.Done()
-	s.daWorkerLoop()
+func (s *Syncer) startP2PListeners() error {
+	if s.headerStore == nil {
+		return errors.New("header store not configured")
+	}
+	if s.dataStore == nil {
+		return errors.New("data store not configured")
+	}
+
+	headerNotifier := s.headerStore.Notifier()
+	if headerNotifier == nil {
+		return errors.New("header notifier not configured")
+	}
+
+	dataNotifier := s.dataStore.Notifier()
+	if dataNotifier == nil {
+		return errors.New("data notifier not configured")
+	}
+
+	s.headerSub = headerNotifier.Subscribe()
+	s.wg.Add(1)
+	go s.consumeNotifierEvents(s.headerSub, syncnotifier.EventTypeHeader)
+
+	s.dataSub = dataNotifier.Subscribe()
+	s.wg.Add(1)
+	go s.consumeNotifierEvents(s.dataSub, syncnotifier.EventTypeData)
+
+	return nil
 }
 
-func (s *Syncer) daWorkerLoop() {
+func (s *Syncer) consumeNotifierEvents(sub *syncnotifier.Subscription, expected syncnotifier.EventType) {
+	defer s.wg.Done()
+
 	if !s.waitForGenesis() {
 		return
 	}
 
-	s.logger.Info().Msg("starting DA worker")
-	defer s.logger.Info().Msg("DA worker stopped")
-
-	nextDARequestAt := &time.Time{}
-	pollInterval := min(10*time.Millisecond, s.config.Node.BlockTime.Duration)
-	if pollInterval <= 0 {
-		pollInterval = 10 * time.Millisecond
-	}
+	logger := s.logger.With().Str("event_type", string(expected)).Logger()
+	logger.Info().Msg("starting P2P notifier worker")
+	defer logger.Info().Msg("P2P notifier worker stopped")
 
 	for {
-		s.tryFetchFromDA(nextDARequestAt)
 		select {
 		case <-s.ctx.Done():
 			return
-		case <-time.After(pollInterval):
+		case evt, ok := <-sub.C:
+			if !ok {
+				return
+			}
+			if evt.Type != expected {
+				logger.Debug().Str("received_type", string(evt.Type)).Msg("ignoring notifier event with unexpected type")
+				continue
+			}
+			s.handleNotifierEvent(evt)
 		}
 	}
 }
 
-func (s *Syncer) p2pWorker() {
+func (s *Syncer) handleNotifierEvent(evt syncnotifier.Event) {
```

Comment:
personal preference: I would inline this method. There is not much semantic added